File: cepi-ios/CEPITelemedicina/Pacientes/PacientesModelo.swift
import Foundation
import Observation

/// Una fila de la lista con lo que la búsqueda y la vista necesitan ya calculado: se arma
/// una vez por carga, no una vez por tecla ni por render.
struct FilaPaciente: Identifiable, Sendable, Hashable {
    let registro: Registro
    let nombre: String
    let cedula: String?
    let iniciales: String
    /// Nombre y cédula sin mayúsculas ni tildes: "jose" encuentra a "José".
    let claveBusqueda: String

    var id: String { registro.id }

    init(_ registro: Registro) {
        self.registro = registro
        let completo = [registro["nombre"], registro["apellidos"]].compactMap { $0 }.joined(separator: " ")
        nombre = completo.isEmpty ? (registro.title ?? "Paciente") : completo
        cedula = registro["cedula"]
        iniciales = nombre.split(separator: " ").prefix(2).compactMap(\.first).map(String.init).joined().uppercased()
        claveBusqueda = FilaPaciente.normalizar("\(nombre) \(cedula ?? "")")
    }

    static func normalizar(_ texto: String) -> String {
        texto.folding(options: [.caseInsensitive, .diacriticInsensitive], locale: Locale(identifier: "es"))
    }
}

@MainActor
@Observable
final class PacientesModelo {
    private(set) var filas: [FilaPaciente] = []
    private(set) var revision: [String: PendienteRevision] = [:]
    private(set) var asignaciones: [String: Asignacion] = [:]
    private(set) var cargado = false
    private(set) var error: String?
    private var porId: [String: FilaPaciente] = [:]
    /// De qué org es lo que hay en `filas`.
    private var organizacion: String?
    /// Cada carga lleva un número; al volver solo se aplica la última. Antes una carga en
    /// curso bloqueaba la siguiente (`guard !cargando`): si se cambiaba de org justo durante
    /// un refresco, la lista de la org nueva no se pedía y la anterior quedaba hasta 20 s.
    private var pedido = 0

    func fila(_ id: String) -> FilaPaciente? { porId[id] }

    /// Pacientes, "revisar" y a cargo en paralelo (la web los pide en serie). Si fallan los
    /// dos accesorios se conserva lo anterior: un error transitorio no debe borrar los
    /// avisos de "revisar".
    func cargar(api: CEPIAPI) async {
        pedido += 1
        let mio = pedido

        async let pacientes = api.pacientes()
        async let cola = try? api.colaRevision()
        async let asignados = try? api.asignaciones()
        do {
            let registros = try await pacientes
            let nuevaCola = await cola
            let nuevasAsignaciones = await asignados
            let nuevasFilas = await Self.preparar(
                registros, revision: nuevaCola ?? revision, asignaciones: nuevasAsignaciones ?? asignaciones
            )
            guard mio == pedido else { return }   // llegó otra carga más nueva (otra org)
            if let nuevaCola { revision = nuevaCola }
            if let nuevasAsignaciones { asignaciones = nuevasAsignaciones }
            filas = nuevasFilas
            porId = Dictionary(nuevasFilas.map { ($0.id, $0) }, uniquingKeysWith: { primera, _ in primera })
            error = nil
            cargado = true
        } catch {
            // En los refrescos periódicos un fallo no pisa la lista; solo se muestra si
            // nunca llegó a cargar.
            guard mio == pedido else { return }
            if !cargado, !Task.isCancelled { self.error = error.localizedDescription }
        }
    }

    /// Otra org es otra lista: se vacía y vuelve a "Cargando pacientes…" en vez de dejar a
    /// la vista, y tocable, la lista de la org anterior mientras llega la nueva.
    func usarOrganizacion(_ org: String?) {
        guard org != organizacion else { return }
        let habiaOtra = organizacion != nil
        organizacion = org
        guard habiaOtra else { return }
        pedido += 1   // lo que llegue de la org anterior se descarta
        filas = []
        porId = [:]
        revision = [:]
        asignaciones = [:]
        cargado = false
        error = nil
    }

    /// Un error de una acción de la lista (borrar), que se muestra donde el resto.
    func mostrarError(_ texto: String) {
        error = texto
    }

    /// Un paciente recién creado aparece ya, sin esperar a la próxima recarga.
    func insertar(_ registro: Registro) {
        guard porId[registro.id] == nil else { return }
        let fila = FilaPaciente(registro)
        filas.insert(fila, at: 0)
        porId[fila.id] = fila
    }

    func estado(_ id: String) -> EstadoFicha { EstadoFicha(asignaciones[id]?.estado) }

    /// Por texto (nombre o cédula, sin tildes) y, si se eligió, por estado de la ficha.
    func filtradas(por busqueda: String, estado: EstadoFicha? = nil) -> [FilaPaciente] {
        let consulta = FilaPaciente.normalizar(busqueda.trimmingCharacters(in: .whitespaces))
        guard !consulta.isEmpty || estado != nil else { return filas }
        return filas.filter { fila in
            (consulta.isEmpty || fila.claveBusqueda.contains(consulta))
                && (estado == nil || self.estado(fila.id) == estado)
        }
    }

    /// Cuántos pacientes hay en cada estado: el filtro los muestra y deshabilita los vacíos.
    func conteoPorEstado() -> [EstadoFicha: Int] {
        filas.reduce(into: [:]) { conteo, fila in conteo[estado(fila.id), default: 0] += 1 }
    }

    /// Fuera del hilo principal: normalizar y ordenar 500 filas no debe costar un frame.
    nonisolated static func preparar(
        _ registros: [Registro], revision: [String: PendienteRevision], asignaciones: [String: Asignacion] = [:]
    ) async -> [FilaPaciente] {
        ordenar(registros.map(FilaPaciente.init), revision: revision, asignaciones: asignaciones)
    }

    /// Por estado de la ficha (el orden de `EstadoFicha`). Dentro de un mismo estado, primero
    /// lo derivado a quien consulta, lo que vence antes arriba (la regla de `filtered` en
    /// ChatList.vue); el resto conserva el orden del servidor.
    nonisolated static func ordenar(
        _ filas: [FilaPaciente], revision: [String: PendienteRevision], asignaciones: [String: Asignacion] = [:]
    ) -> [FilaPaciente] {
        let vence = revision.mapValues { Fechas.iso($0.vence) ?? .distantFuture }
        let conEstado = filas.enumerated().map { indice, fila in
            (fila: fila, orden: indice, estado: EstadoFicha(asignaciones[fila.id]?.estado).rawValue)
        }
        return conEstado.sorted { a, b in
            if a.estado != b.estado { return a.estado < b.estado }
            switch (vence[a.fila.id], vence[b.fila.id]) {
            case (.some, .none):
                return true
            case (.none, .some):
                return false
            case let (.some(va), .some(vb)) where va != vb:
                return va < vb
            default:
                return a.orden < b.orden
            }
        }.map(\.fila)
    }
}

File: cepi-android/app/src/main/java/ec/cepi/telemedicina/pacientes/PacientesModelo.kt
package ec.cepi.telemedicina.pacientes

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import ec.cepi.telemedicina.api.ApiError
import ec.cepi.telemedicina.api.Asignacion
import ec.cepi.telemedicina.api.CepiApi
import ec.cepi.telemedicina.api.Fechas
import ec.cepi.telemedicina.api.PendienteRevision
import ec.cepi.telemedicina.api.Registro
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.supervisorScope
import kotlinx.coroutines.withContext
import java.text.Normalizer
import java.time.Instant
import java.util.Locale
import java.util.concurrent.atomic.AtomicInteger

/**
 * Una fila de la lista con lo que la búsqueda y la vista necesitan ya calculado: se arma una
 * vez por carga, no una vez por tecla ni por recomposición.
 */
class FilaPaciente(val registro: Registro) {
    val id: String = registro.id
    val nombre: String
    val cedula: String? = registro["cedula"]
    val iniciales: String
    /** Nombre y cédula sin mayúsculas ni tildes: "jose" encuentra a "José". */
    val claveBusqueda: String

    init {
        val completo = listOfNotNull(registro["nombre"], registro["apellidos"]).joinToString(" ")
        nombre = completo.ifEmpty { registro.title ?: "Paciente" }
        iniciales = nombre.split(' ').filter { it.isNotEmpty() }.take(2).map { it.first() }.joinToString("").uppercase()
        claveBusqueda = normalizar("$nombre ${cedula.orEmpty()}")
    }

    companion object {
        private val marcas = "\\p{Mn}+".toRegex()

        fun normalizar(texto: String): String =
            Normalizer.normalize(texto, Normalizer.Form.NFD).replace(marcas, "").lowercase(Locale.ROOT)
    }
}

/** Lista de pacientes, "revisar" y a cargo. Espejo de `PacientesModelo.swift`. */
class PacientesModelo {
    var filas: List<FilaPaciente> by mutableStateOf(emptyList())
        private set
    var revision: Map<String, PendienteRevision> by mutableStateOf(emptyMap())
        private set
    var asignaciones: Map<String, Asignacion> by mutableStateOf(emptyMap())
        private set
    var cargado: Boolean by mutableStateOf(false)
        private set
    var error: String? by mutableStateOf(null)
        private set

    private var porId: Map<String, FilaPaciente> = emptyMap()
    /** De qué org es lo que hay en `filas`. */
    private var organizacion: String? = null
    /**
     * Cada carga lleva un número; al volver solo se aplica la última. Si se cambia de org
     * justo durante un refresco, lo que llegue de la org anterior se descarta.
     */
    private val pedido = AtomicInteger(0)

    fun fila(id: String): FilaPaciente? = porId[id]

    /**
     * Pacientes, "revisar" y a cargo en paralelo (la web los pide en serie). Si fallan los dos
     * accesorios se conserva lo anterior: un error transitorio no debe borrar los avisos de
     * "revisar".
     */
    suspend fun cargar(api: CepiApi) {
        val mio = pedido.incrementAndGet()
        try {
            val (registros, nuevaCola, nuevasAsignaciones) = supervisorScope {
                val pacientes = async { api.pacientes() }
                val cola = async { intentar { api.colaRevision() } }
                val asignados = async { intentar { api.asignaciones() } }
                Triple(pacientes.await(), cola.await(), asignados.await())
            }
            // Fuera del hilo principal: normalizar y ordenar 500 filas no debe costar un frame.
            val nuevasFilas = withContext(Dispatchers.Default) { preparar(registros, nuevaCola ?: revision) }
            if (mio != pedido.get()) return   // llegó otra carga más nueva (otra org)
            nuevaCola?.let { revision = it }
            nuevasAsignaciones?.let { asignaciones = it }
            filas = nuevasFilas
            porId = nuevasFilas.associateBy { it.id }
            error = null
            cargado = true
        } catch (e: ApiError) {
            // En los refrescos periódicos un fallo no pisa la lista; solo se muestra si nunca
            // llegó a cargar.
            if (mio != pedido.get()) return
            if (!cargado) error = e.mensaje
        }
    }

    /**
     * Otra org es otra lista: se vacía y vuelve a "Cargando pacientes…" en vez de dejar a la
     * vista, y tocable, la lista de la org anterior mientras llega la nueva.
     */
    fun usarOrganizacion(org: String?) {
        if (org == organizacion) return
        val habiaOtra = organizacion != null
        organizacion = org
        if (!habiaOtra) return
        pedido.incrementAndGet()   // lo que llegue de la org anterior se descarta
        filas = emptyList()
        porId = emptyMap()
        revision = emptyMap()
        asignaciones = emptyMap()
        cargado = false
        error = null
    }

    /** Un paciente recién creado aparece ya, sin esperar a la próxima recarga. */
    fun insertar(registro: Registro) {
        if (registro.id in porId) return
        val fila = FilaPaciente(registro)
        filas = listOf(fila) + filas
        porId = porId + (fila.id to fila)
    }

    fun filtradas(busqueda: String): List<FilaPaciente> {
        val consulta = FilaPaciente.normalizar(busqueda.trim())
        if (consulta.isEmpty()) return filas
        return filas.filter { consulta in it.claveBusqueda }
    }

    companion object {
        fun preparar(registros: List<Registro>, revision: Map<String, PendienteRevision>): List<FilaPaciente> =
            ordenar(registros.map(::FilaPaciente), revision)

        /**
         * Primero lo derivado a quien consulta, lo que vence antes arriba; el resto conserva el
         * orden del servidor (`sortedWith` es estable). Misma regla que `filtered` en ChatList.vue.
         */
        fun ordenar(filas: List<FilaPaciente>, revision: Map<String, PendienteRevision>): List<FilaPaciente> {
            val vence = revision.mapValues { Fechas.iso(it.value.vence) ?: Instant.MAX }
            return filas.sortedWith(
                compareBy<FilaPaciente> { if (it.id in vence) 0 else 1 }.thenBy { vence[it.id] ?: Instant.MAX },
            )
        }

        private suspend fun <T> intentar(bloque: suspend () -> T): T? = try {
            bloque()
        } catch (_: ApiError) {
            null
        }
    }
}
